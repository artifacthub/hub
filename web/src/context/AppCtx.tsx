import isUndefined from 'lodash/isUndefined';
import React, { createContext, useContext, useEffect, useReducer } from 'react';

import { API } from '../api';
import useSystemThemeMode from '../hooks/useSystemThemeMode';
import { Prefs, Profile, ThemePrefs, UserFullName } from '../types';
import cleanLoginUrlParams from '../utils/cleanLoginUrlParams';
import detectActiveThemeMode from '../utils/detectActiveThemeMode';
import history from '../utils/history';
import isControlPanelSectionAvailable from '../utils/isControlPanelSectionAvailable';
import lsStorage from '../utils/localStoragePreferences';

interface AppState {
  user: Profile | null | undefined;
  prefs: Prefs;
}

interface Props {
  children: JSX.Element;
}

const initialState: AppState = {
  user: undefined,
  prefs: lsStorage.getPrefs(),
};

type Action =
  | { type: 'signIn'; profile: Profile }
  | { type: 'signOut' }
  | { type: 'unselectOrg' }
  | { type: 'updateUser'; user: UserFullName }
  | { type: 'updateOrg'; name: string }
  | { type: 'updateLimit'; limit: number }
  | { type: 'updateTheme'; theme: string }
  | { type: 'updateEffectiveTheme'; theme: string }
  | { type: 'enabledDisplayedNotifications'; enabled: boolean }
  | { type: 'addNewDisplayedNotification'; id: string };

export const AppCtx = createContext<{
  ctx: AppState;
  dispatch: React.Dispatch<any>;
}>({
  ctx: initialState,
  dispatch: () => null,
});

export function signOut() {
  return { type: 'signOut' };
}

export function unselectOrg() {
  return { type: 'unselectOrg' };
}

export function updateOrg(name: string) {
  return { type: 'updateOrg', name };
}

export function updateUser(user: UserFullName) {
  return { type: 'updateUser', user };
}

export function updateLimit(limit: number) {
  return { type: 'updateLimit', limit };
}

export function updateTheme(theme: string) {
  return { type: 'updateTheme', theme };
}

export function updateEffectiveTheme(theme: string) {
  return { type: 'updateTheme', theme };
}

export function enabledDisplayedNotifications(enabled: boolean) {
  return { type: 'enabledDisplayedNotifications', enabled };
}

export function addNewDisplayedNotification(id: string) {
  return { type: 'addNewDisplayedNotification', id };
}

export async function refreshUserProfile(dispatch: React.Dispatch<any>, redirectUrl?: string) {
  try {
    const profile: Profile = await API.getUserProfile();
    dispatch({ type: 'signIn', profile });
    const currentUrl = `${window.location.pathname}${
      window.location.search !== '' ? `?${cleanLoginUrlParams(window.location.search)}` : ''
    }`;
    if (!isUndefined(redirectUrl)) {
      if (redirectUrl === currentUrl) {
        history.replace(redirectUrl);
      } else {
        // Redirect to correct route when neccessary
        history.push(redirectUrl);
      }
    }
  } catch (err) {
    dispatch({ type: 'signOut' });
    if (err.message === 'invalid session') {
      history.push(
        `${window.location.pathname}${
          window.location.search === '' ? '?' : `${window.location.search}&`
        }modal=login&redirect=${encodeURIComponent(`${window.location.pathname}${window.location.search}`)}`
      );
    }
  }
}

function redirectToControlPanel(context: 'user' | 'org') {
  if (history.location.pathname.startsWith('/control-panel')) {
    const sections = history.location.pathname.split('/');
    if (!isControlPanelSectionAvailable(context, sections[2], sections[3])) {
      history.push('/control-panel/repositories');
    }
  }
}

function updateSelectedOrg(currentPrefs: Prefs, name?: string): Prefs {
  return {
    ...currentPrefs,
    controlPanel: {
      ...currentPrefs.controlPanel,
      selectedOrg: name,
    },
  };
}

export function updateActiveStyleSheet(current: string) {
  document.getElementsByTagName('html')[0].setAttribute('data-theme', current);
}

function getCurrentSystemActiveTheme(prefs: ThemePrefs): ThemePrefs {
  if (
    prefs.configured === 'automatic' &&
    (isUndefined(prefs.effective) || detectActiveThemeMode() !== prefs.effective)
  ) {
    return {
      ...prefs,
      effective: detectActiveThemeMode(),
    };
  } else {
    return prefs;
  }
}

export function appReducer(state: AppState, action: Action) {
  let prefs;
  switch (action.type) {
    case 'signIn':
      prefs = lsStorage.getPrefs(action.profile.alias);
      const userPrefs = { ...prefs, theme: getCurrentSystemActiveTheme(prefs.theme) };
      updateActiveStyleSheet(userPrefs.theme.effective);
      lsStorage.setPrefs(userPrefs, action.profile.alias);
      lsStorage.setActiveProfile(action.profile.alias);
      return {
        user: action.profile,
        prefs: userPrefs,
      };

    case 'unselectOrg':
      prefs = updateSelectedOrg(state.prefs);
      lsStorage.setPrefs(prefs, state.user!.alias);
      redirectToControlPanel('user');
      return {
        ...state,
        prefs: prefs,
      };

    case 'signOut':
      prefs = lsStorage.getPrefs();
      const guestPrefs = { ...prefs, theme: getCurrentSystemActiveTheme(prefs.theme) };
      lsStorage.setPrefs(guestPrefs);
      lsStorage.setActiveProfile();
      updateActiveStyleSheet(guestPrefs.theme.effective);
      return { user: null, prefs: guestPrefs };

    case 'updateOrg':
      prefs = updateSelectedOrg(state.prefs, action.name);
      lsStorage.setPrefs(prefs, state.user!.alias);
      if (isUndefined(state.prefs.controlPanel.selectedOrg) || action.name !== state.prefs.controlPanel.selectedOrg) {
        redirectToControlPanel('org');
      }
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateLimit':
      prefs = {
        ...state.prefs,
        search: {
          ...state.prefs.search,
          limit: action.limit,
        },
      };
      lsStorage.setPrefs(prefs, state.user ? state.user.alias : undefined);
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateTheme':
      const effective = action.theme === 'automatic' ? detectActiveThemeMode() : action.theme;
      prefs = {
        ...state.prefs,
        theme: {
          configured: action.theme,
          effective: effective,
        },
      };

      lsStorage.setPrefs(prefs, state.user ? state.user.alias : undefined);
      updateActiveStyleSheet(effective);
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateEffectiveTheme':
      prefs = {
        ...state.prefs,
        theme: {
          ...state.prefs.theme,
          effective: action.theme,
        },
      };
      lsStorage.setPrefs(prefs, state.user ? state.user.alias : undefined);
      updateActiveStyleSheet(action.theme);
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateUser':
      lsStorage.updateAlias(state.user!.alias, action.user.alias);
      lsStorage.setActiveProfile(action.user.alias);
      return {
        ...state,
        user: {
          ...state.user!,
          ...action.user,
        },
      };

    case 'enabledDisplayedNotifications':
      prefs = {
        ...state.prefs,
        notifications: {
          ...state.prefs.notifications,
          enabled: action.enabled,
        },
      };
      lsStorage.setPrefs(prefs, state.user ? state.user.alias : undefined);
      return {
        ...state,
        prefs: prefs,
      };

    case 'addNewDisplayedNotification':
      prefs = {
        ...state.prefs,
        notifications: {
          ...state.prefs.notifications,
          displayed: [...state.prefs.notifications.displayed, action.id],
          lastDisplayedTime: Date.now(),
        },
      };
      lsStorage.setPrefs(prefs, state.user ? state.user.alias : undefined);
      return {
        ...state,
        prefs: prefs,
      };

    default:
      return { ...state };
  }
}

function AppCtxProvider(props: Props) {
  const [ctx, dispatch] = useReducer(appReducer, initialState);

  useEffect(() => {
    refreshUserProfile(dispatch);
  }, []);

  useSystemThemeMode(ctx.prefs.theme.configured === 'automatic', dispatch);

  return <AppCtx.Provider value={{ ctx, dispatch }}>{props.children}</AppCtx.Provider>;
}

function useAppCtx() {
  return useContext(AppCtx);
}

export { AppCtxProvider, useAppCtx };
