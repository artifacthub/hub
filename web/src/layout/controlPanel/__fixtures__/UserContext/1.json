[
  {
    "name": "test",
    "confirmed": true,
    "membersCount": 2
  }
]
