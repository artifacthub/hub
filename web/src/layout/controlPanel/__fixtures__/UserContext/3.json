[
  {
    "name": "test",
    "confirmed": true,
    "membersCount": 2
  },
  {
    "name": "test1",
    "displayName": "Test name",
    "description": "This is a org description",
    "homeUrl": "http://test.com",
    "confirmed": true,
    "membersCount": 1
  }
]
