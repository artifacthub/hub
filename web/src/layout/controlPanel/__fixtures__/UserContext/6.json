[
  {
    "name": "test",
    "displayName": null,
    "description": null,
    "homeUrl": null,
    "logoImageId": null,
    "confirmed": true,
    "membersCount": 2
  },
  {
    "name": "test1",
    "displayName": "Test name",
    "description": "This is a org description",
    "homeUrl": "http://test.com",
    "logoImageId": null,
    "confirmed": true,
    "membersCount": 1
  }
]
