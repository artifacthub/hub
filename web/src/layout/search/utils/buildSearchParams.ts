import { Filters } from '../../../types';
import isNull from 'lodash/isNull';

const SPECIAL_KEYS = ['text', 'page'];

interface F {
  [key: string]: string[];
}

export default (query: string): Filters => {
  const p = new URLSearchParams(query);
  let f: F = {};

  p.forEach((value, key) => {
    if (!SPECIAL_KEYS.includes(key)) {
      const values = f[key] || [];
      values.push(value);
      f[key] = values;
    }
  });

  return {
    text: p.has('text') ? p.get('text')! : undefined,
    f: { ...f },
    pageNumber: p.has('page') && !isNull(p.get('page')) ? parseInt(p.get('page')!) : 1,
  };
}
