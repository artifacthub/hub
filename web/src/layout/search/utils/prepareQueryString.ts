import isUndefined from 'lodash/isUndefined';
import { Filters } from '../../../types';

export default (filters: Filters): string => {
  const p = new URLSearchParams();
  p.set('page', filters.pageNumber.toString());
  if (!isUndefined(filters.text)) {
    p.set('text', filters.text);
  }
  if (!isUndefined(filters.f)) {
    Object.keys(filters.f).forEach((filterId: string) => {
      return filters.f[filterId].forEach((id: string) => {
        p.append(filterId, id);
      });
    });
  }
  const result = p.toString();
  return result === '' ? '' : `?${result}`;
}
