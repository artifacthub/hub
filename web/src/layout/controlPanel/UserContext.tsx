import classnames from 'classnames';
import isString from 'lodash/isString';
import isUndefined from 'lodash/isUndefined';
import React, { useContext, useEffect, useRef, useState } from 'react';
import { FaCaretDown, FaUser } from 'react-icons/fa';
import { GoCheck } from 'react-icons/go';
import { MdBusiness } from 'react-icons/md';

import { API } from '../../api';
import { AppCtx, unselectOrg, updateOrg } from '../../context/AppCtx';
import useOutsideClick from '../../hooks/useOutsideClick';
import { ErrorKind, Organization } from '../../types';
import authorizer from '../../utils/authorizer';
import styles from './UserContext.module.css';

const UserContext = () => {
  const { ctx, dispatch } = useContext(AppCtx);
  const [organizations, setOrganizations] = useState<Organization[] | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const alias = ctx.user!.alias;
  const [openStatus, setOpenStatus] = useState(false);
  const ref = useRef(null);
  useOutsideClick([ref], openStatus, () => setOpenStatus(false));

  const handleChange = (value: string | Organization): void => {
    if (isString(value)) {
      authorizer.updateCtx();
      dispatch(unselectOrg());
    } else {
      authorizer.updateCtx(value.name);
      dispatch(updateOrg(value.name));
    }
    setOpenStatus(false);
  };

  async function fetchOrganizations() {
    try {
      setIsLoading(true);
      const allOrganizations = await API.getUserOrganizations();
      const confirmedOrganizations = allOrganizations.filter((org: Organization) => org.confirmed);
      if (ctx.prefs.controlPanel.selectedOrg) {
        const selectedOrg = confirmedOrganizations.find(
          (org: Organization) => org.name === ctx.prefs.controlPanel.selectedOrg
        );
        if (isUndefined(selectedOrg)) {
          dispatch(unselectOrg());
        } else {
          authorizer.updateCtx(ctx.prefs.controlPanel.selectedOrg);
        }
      }
      setOrganizations(confirmedOrganizations);
      setIsLoading(false);
    } catch (err) {
      setIsLoading(false);
      if (err.kind !== ErrorKind.Unauthorized) {
        setOrganizations([]);
      }
    }
  }

  useEffect(() => {
    fetchOrganizations();
    authorizer.init(ctx.prefs.controlPanel.selectedOrg);
  }, []); /* eslint-disable-line react-hooks/exhaustive-deps */

  return (
    <div className={`position-relative ${styles.ctxWrapper}`}>
      <div className="d-flex flex-column">
        <small className={`text-uppercase text-muted ${styles.legendCtx}`}>Control panel context</small>
        <div className="d-flex flex-row align-items-center">
          <button
            data-testid="ctxBtn"
            className={`btn btn-primary badge-pill btn-sm pr-3 position-relative ${styles.ctxBtn}`}
            type="button"
            onClick={() => {
              fetchOrganizations();
              setOpenStatus(true);
            }}
          >
            <div className="d-flex flex-row align-items-center">
              {!isUndefined(ctx.prefs.controlPanel.selectedOrg) ? (
                <>
                  <div className={`badge badge-light badge-pill mr-2 p-0 ${styles.badgeIcon}`}>
                    <MdBusiness />
                  </div>
                  <div className="flex-grow-1 text-left mr-1 text-truncate">{ctx.prefs.controlPanel.selectedOrg}</div>
                </>
              ) : (
                <>
                  <div className={`badge badge-light badge-pill mr-2 p-0 ${styles.badgeIcon}`}>
                    <FaUser />
                  </div>
                  <div className="flex-grow-1 text-left mr-1 text-truncate">{alias}</div>
                </>
              )}
            </div>

            <div className={`position-absolute textLight ${styles.caret}`}>
              <FaCaretDown />
            </div>
          </button>
        </div>
        {isLoading && (
          <div className={`position-absolute text-secondary ${styles.loading}`} role="status">
            <span className="spinner-border spinner-border-sm" />
          </div>
        )}
      </div>

      <div
        ref={ref}
        data-testid="ctxDropdown"
        className={classnames('dropdown-menu dropdown-menu-right', styles.dropdown, { show: openStatus })}
      >
        <div className={`arrow ${styles.arrow}`} />

        <button data-testid="userCtxBtn" className="dropdown-item mw-100" onClick={() => handleChange(alias)}>
          <div className="d-flex flex-row align-items-center text-truncate">
            <FaUser className={`mr-2 ${styles.icon}`} />
            <div className="flex-grow-1 text-truncate">{alias}</div>
            {isUndefined(ctx.prefs.controlPanel.selectedOrg) && (
              <GoCheck className={`ml-2 text-success ${styles.icon}`} />
            )}
          </div>
        </button>
        {organizations && (
          <>
            {organizations.map((org: Organization) => (
              <button
                data-testid="orgCtxBtn"
                key={`opt_${org.name}`}
                className="dropdown-item"
                onClick={() => handleChange(org)}
              >
                <div className="d-flex flex-row align-items-center text-truncate">
                  <MdBusiness className={`mr-2 ${styles.icon}`} />
                  <div className="flex-grow-1 text-truncate">{org.name}</div>
                  {!isUndefined(ctx.prefs.controlPanel.selectedOrg) &&
                    org.name === ctx.prefs.controlPanel.selectedOrg && (
                      <GoCheck className={`ml-2 text-success ${styles.icon}`} />
                    )}
                </div>
              </button>
            ))}
          </>
        )}
      </div>
    </div>
  );
};

export default UserContext;
