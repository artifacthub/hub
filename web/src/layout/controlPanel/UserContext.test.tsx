import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { mocked } from 'ts-jest/utils';

import { API } from '../../api';
import { AppCtx } from '../../context/AppCtx';
import { Organization } from '../../types';
import UserContext from './UserContext';
jest.mock('../../api');

const getMockOrgs = (fixtureId: string): Organization[] => {
  return require(`./__fixtures__/UserContext/${fixtureId}.json`) as Organization[];
};

const mockCtx = {
  user: { alias: 'userAlias', email: 'jsmith@email.com' },
  prefs: {
    controlPanel: {},
    search: { limit: 60 },
    theme: {
      configured: 'light',
      effective: 'light',
    },
    notifications: {
      lastDisplayedTime: null,
      enabled: true,
      displayed: [],
    },
  },
};

const mockOrgCtx = {
  user: { alias: 'userAlias', email: 'jsmith@email.com' },
  prefs: {
    controlPanel: {
      selectedOrg: 'test',
    },
    search: { limit: 60 },
    theme: {
      configured: 'light',
      effective: 'light',
    },
    notifications: {
      lastDisplayedTime: null,
      enabled: true,
      displayed: [],
    },
  },
};

const mockOrgCtx1 = {
  user: { alias: 'userAlias', email: 'jsmith@email.com' },
  prefs: {
    controlPanel: {
      selectedOrg: 'org',
    },
    search: { limit: 60 },
    theme: {
      configured: 'light',
      effective: 'light',
    },
    notifications: {
      lastDisplayedTime: null,
      enabled: true,
      displayed: [],
    },
  },
};

const mockDispatch = jest.fn();

describe('UserContext', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', async () => {
    const mockOrgs = getMockOrgs('1');
    mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

    const result = render(
      <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
        <Router>
          <UserContext />
        </Router>
      </AppCtx.Provider>
    );

    await waitFor(() => {
      expect(API.getUserOrganizations).toHaveBeenCalledTimes(1);
      expect(result.asFragment()).toMatchSnapshot();
    });
  });

  describe('Render', () => {
    it('renders component', async () => {
      const mockOrgs = getMockOrgs('2');
      mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

      render(
        <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
          <Router>
            <UserContext />
          </Router>
        </AppCtx.Provider>
      );

      await waitFor(() => {
        expect(API.getUserOrganizations).toHaveBeenCalledTimes(1);
      });
    });

    it('displays spinner to get organizations', async () => {
      const mockOrgs = getMockOrgs('3');
      mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

      const { getByRole } = render(
        <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
          <Router>
            <UserContext />
          </Router>
        </AppCtx.Provider>
      );

      await waitFor(() => {
        expect(getByRole('status')).toBeTruthy();
      });
    });

    it('displays dropdown with ctx', async () => {
      const mockOrgs = getMockOrgs('4');
      mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

      const { getByTestId, getAllByTestId } = render(
        <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
          <Router>
            <UserContext />
          </Router>
        </AppCtx.Provider>
      );

      await waitFor(() => {
        expect(API.getUserOrganizations).toHaveBeenCalledTimes(1);
      });

      const ctxBtn = getByTestId('ctxBtn');
      const ctxDropdown = getByTestId('ctxDropdown');

      expect(ctxBtn).toBeInTheDocument();
      expect(ctxDropdown).toBeInTheDocument();
      expect(ctxDropdown).not.toHaveClass('show');

      fireEvent.click(ctxBtn);

      expect(ctxDropdown).toHaveClass('show');

      await waitFor(() => {
        expect(API.getUserOrganizations).toHaveBeenCalledTimes(2);
      });

      expect(getByTestId('userCtxBtn')).toBeInTheDocument();
      expect(getAllByTestId('orgCtxBtn')).toHaveLength(mockOrgs.length);
    });

    it('renders only user ctx when no orgs', async () => {
      const mockOrgs = getMockOrgs('5');
      mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

      const { getByTestId, queryAllByTestId } = render(
        <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
          <Router>
            <UserContext />
          </Router>
        </AppCtx.Provider>
      );

      await waitFor(() => {
        expect(API.getUserOrganizations).toHaveBeenCalledTimes(1);
        expect(getByTestId('userCtxBtn')).toBeInTheDocument();
        expect(queryAllByTestId('orgCtxBtn')).toHaveLength(0);
      });
    });

    it('calls updateOrg when org ctx button is clicked', async () => {
      const mockOrgs = getMockOrgs('4');
      mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

      const { getByTestId, getAllByTestId } = render(
        <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
          <Router>
            <UserContext />
          </Router>
        </AppCtx.Provider>
      );

      await waitFor(() => {
        expect(API.getUserOrganizations).toHaveBeenCalledTimes(1);
      });

      const ctxBtn = getByTestId('ctxBtn');
      const ctxDropdown = getByTestId('ctxDropdown');

      expect(ctxBtn).toBeInTheDocument();
      expect(ctxDropdown).toBeInTheDocument();
      expect(ctxDropdown).not.toHaveClass('show');

      fireEvent.click(ctxBtn);

      expect(ctxDropdown).toHaveClass('show');

      expect(API.getUserOrganizations).toHaveBeenCalledTimes(2);

      expect(getByTestId('userCtxBtn')).toBeInTheDocument();

      const orgBtns = getAllByTestId('orgCtxBtn');
      expect(orgBtns).toHaveLength(mockOrgs.length);

      fireEvent.click(orgBtns[0]);

      await waitFor(() => {
        expect(mockDispatch).toHaveBeenCalledTimes(1);
        expect(mockDispatch).toHaveBeenCalledWith({ name: 'test', type: 'updateOrg' });
      });

      expect(ctxDropdown).not.toHaveClass('show');
    });

    it('calls unselectOrg when user ctx button is clicked', async () => {
      const mockOrgs = getMockOrgs('4');
      mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

      const { getByTestId } = render(
        <AppCtx.Provider value={{ ctx: mockOrgCtx, dispatch: mockDispatch }}>
          <Router>
            <UserContext />
          </Router>
        </AppCtx.Provider>
      );

      await waitFor(() => {
        expect(API.getUserOrganizations).toHaveBeenCalledTimes(1);
      });

      const ctxBtn = getByTestId('ctxBtn');
      fireEvent.click(ctxBtn);

      expect(API.getUserOrganizations).toHaveBeenCalledTimes(2);

      const userCtxBtn = getByTestId('userCtxBtn');
      fireEvent.click(userCtxBtn);

      await waitFor(() => {
        expect(mockDispatch).toHaveBeenCalledTimes(1);
        expect(mockDispatch).toHaveBeenCalledWith({ type: 'unselectOrg' });
      });
    });

    it('calls unselectOrg when selectedOrg is not in the list', async () => {
      const mockOrgs = getMockOrgs('4');
      mocked(API).getUserOrganizations.mockResolvedValue(mockOrgs);

      render(
        <AppCtx.Provider value={{ ctx: mockOrgCtx1, dispatch: mockDispatch }}>
          <Router>
            <UserContext />
          </Router>
        </AppCtx.Provider>
      );

      await waitFor(() => {
        expect(API.getUserOrganizations).toHaveBeenCalledTimes(1);
      });

      await waitFor(() => {
        expect(mockDispatch).toHaveBeenCalledTimes(1);
        expect(mockDispatch).toHaveBeenCalledWith({ type: 'unselectOrg' });
      });
    });
  });
});
