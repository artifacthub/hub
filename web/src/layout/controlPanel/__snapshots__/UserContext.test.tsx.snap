// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserContext creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="position-relative ctxWrapper"
  >
    <div
      class="d-flex flex-column"
    >
      <small
        class="text-uppercase text-muted legendCtx"
      >
        Control panel context
      </small>
      <div
        class="d-flex flex-row align-items-center"
      >
        <button
          class="btn btn-primary badge-pill btn-sm pr-3 position-relative ctxBtn"
          data-testid="ctxBtn"
          type="button"
        >
          <div
            class="d-flex flex-row align-items-center"
          >
            <div
              class="badge badge-light badge-pill mr-2 p-0 badgeIcon"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                />
              </svg>
            </div>
            <div
              class="flex-grow-1 text-left mr-1 text-truncate"
            >
              userAlias
            </div>
          </div>
          <div
            class="position-absolute textLight caret"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        class="position-absolute text-secondary loading"
        role="status"
      >
        <span
          class="spinner-border spinner-border-sm"
        />
      </div>
    </div>
    <div
      class="dropdown-menu dropdown-menu-right dropdown"
      data-testid="ctxDropdown"
    >
      <div
        class="arrow arrow"
      />
      <button
        class="dropdown-item mw-100"
        data-testid="userCtxBtn"
      >
        <div
          class="d-flex flex-row align-items-center text-truncate"
        >
          <svg
            class="mr-2 icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
          <div
            class="flex-grow-1 text-truncate"
          >
            userAlias
          </div>
          <svg
            class="ml-2 text-success icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 12 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 5l-8 8-4-4 1.5-1.5L4 10l6.5-6.5L12 5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
